### PR TITLE
Add a hyper parameter tuning and tensorboard feature to the nn trai…

### PR DESCRIPTION
**Pull request by yyajima21**
# Description
This PR contains several changes to the existing nn training code.

This PR does the following:
- Added a hyper-parameter tuning code
- Refactored the training code to enable the hyper-parameter tuning code
- Added a tensor board feature to visualize training process
- Added print statements to see the training timer and overall training loss

This PR Fixes the following issues #565 #549

# Testing steps (If relevant)
## Test Case 1
1. Please download the training and validation dataset from the Robojackets cloud
[https://cloud.robojackets.org/apps/files/?dir=/IGVC/Software/dataset/IGVC&fileid=338649](url)
2. Open your google colab and load the notebook named "igvc-unet"
3. Modify your cfg file based on your file path
4. Run the following code `python train.py --epoch 100 --save_model --save_interval 10`
5. Check your training result by `tensorbaord --logdir="your runs file path"`

Expectation: This PR allows our current training model to visualize and compare trained models with different hyper-parameters. This PR also refactors the training code. 

# Self Checklist
- [x] I have formatted my code using `make format`
- [x] I have tested that the new behaviour works

### CODE REVIEW

**Comment by suhacker1** [comment]
My initial suggestion is to try using [Python Black](https://pypi.org/project/black/) to get rid of the PEP 8 violations in your code

File: igvc_perception/src/train_eval/IGVCDataset.py [hunk 1]
```diff
@@ -49,11 +50,26 @@ def __getitem__(self, index):
...
+            target = cv2.flip(target,1)
+
+        #Access to red color only (OpenCV: BGR) This is because the lines are categorized as red color
+        # And we are interested in lines for the line detection.
+        target = target[:,:,2]
         img = cv2.resize(img, (self.im_size[1],self.im_size[2]))
         target = cv2.resize(target, (self.im_size[1], self.im_size[2]))
         target[target != 0] = 255
+        #cv2.imshow("original image", img)
+        #cv2.imshow("target", target) #Check if the image is collectlly displayed as a binary image
```
Comment (suhacker1) [inline igvc_perception/src/train_eval/IGVCDataset.py:71]: I think you should uncomment these lines, delete them, or add some kind of logic to thoroughly ensure that it is displayed as a binary image.

Comment (suhacker1) [reply igvc_perception/src/train_eval/IGVCDataset.py:71]: Another possibility to just space it out and better distinguish it as a checking mechanism.

File: igvc_perception/src/train_eval/train.py [hunk 2]
```diff
@@ -2,167 +2,98 @@
...
+            loss.backward()
+            optimizer.step()
+
+            if args.vis and batch_idx % args.log_interval == 0 and images.shape[0] == 1:
+                cv2.imshow('output: ', outputs.cpu().data.numpy()[0][0])
+                cv2.imshow('target: ', targets.cpu().data.numpy()[0][0])
+                cv2.waitKey(10)
+
+            if batch_idx % args.log_interval == 0:
+                val_loss, val_acc = evaluate('val', n_batches=80)
```
Comment (suhacker1) [inline igvc_perception/src/train_eval/train.py:62]: The number of batches is hard coded here. I suggest turning it into a variable.

File: igvc_perception/src/train_eval/IGVCDataset.py (hunk 1 above)
Comment (yyajima21) [reply igvc_perception/src/train_eval/IGVCDataset.py]: Thanks @suhacker1, I updated my code based on your feedback!

**Comment by yyajima21** [comment]
@suhacker1, thank you for your review. I included a google colab notebook that contains all major library requirements. Let me know if this google colab notebook works.
